File: static/java/TeamPhoto2.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

public class TeamPhoto2 {
  // @include
  public static class GraphVertex {
    public List<GraphVertex> edges = new ArrayList<>();
    // Set maxDistance = 0 to indicate unvisited vertex.
    public int maxDistance = 0;
  }

  public static int findLargestNumberTeams(List<GraphVertex> G) {
    return findLongestPath(buildTopologicalOrdering(G));
  }

  private static Deque<GraphVertex> buildTopologicalOrdering(
      List<GraphVertex> G) {
    Deque<GraphVertex> orderedVertices = new LinkedList<>();
    for (GraphVertex g : G) {
      if (g.maxDistance == 0) {
        DFS(g, orderedVertices);
      }
    }
    return orderedVertices;
  }

  private static int findLongestPath(Deque<GraphVertex> orderedVertices) {
    int maxDistance = 0;
    while (!orderedVertices.isEmpty()) {
      GraphVertex u = orderedVertices.peekFirst();
      maxDistance = Math.max(maxDistance, u.maxDistance);
      for (GraphVertex v : u.edges) {
        v.maxDistance = Math.max(v.maxDistance, u.maxDistance + 1);
      }
      orderedVertices.removeFirst();
    }
    return maxDistance;
  }

  private static void DFS(GraphVertex curr,
                          Deque<GraphVertex> orderedVertices) {
    curr.maxDistance = 1;
    for (GraphVertex vertex : curr.edges) {
      if (vertex.maxDistance == 0) {
        DFS(vertex, orderedVertices);
      }
    }
    orderedVertices.addFirst(curr);
  }
  // @exclude

  public static void main(String[] args) {
    List<GraphVertex> G = new ArrayList<>(3);
    for (int i = 0; i < 3; i++) {
      G.add(new GraphVertex());
    }
    G.get(0).edges.add(G.get(2));
    G.get(1).edges.add(G.get(2));
    assert(2 == findLargestNumberTeams(G));
  }
}

File: static/python/team_photo_2.py
# @include
class GraphVertex:
    def __init__(self):
        self.edges = []
        self.max_distance = 0
# @exclude

    def __repr__(self):
        return ('*' if self.max_distance else '') + '(%d)%d(%s)' % (
            self.max_distance, id(self), ','.join(
                str(id(x)) for x in self.edges))


# @include


def find_largest_number_teams(G):
    def build_topological_ordering():
        def dfs(cur):
            cur.max_distance = 1
            for next in cur.edges:
                if not next.max_distance:
                    dfs(next)
            vertex_order.append(cur)

        vertex_order = []
        for g in G:
            if not g.max_distance:
                dfs(g)
        return vertex_order

    def find_longest_path(vertex_order):
        max_distance = 0
        while vertex_order:
            u = vertex_order.pop()
            max_distance = max(max_distance, u.max_distance)
            for v in u.edges:
                v.max_distance = max(v.max_distance, u.max_distance + 1)
        return max_distance

    return find_longest_path(build_topological_ordering())


# @exclude


def main():
    G = [GraphVertex() for _ in range(3)]
    G[0].edges.append(G[2])
    G[1].edges.append(G[2])
    assert 2 == find_largest_number_teams(G)


if __name__ == '__main__':
    main()
